service/storage_service.cc

```diff
@@ -596,41 +611,60 @@ future<storage_service::nodes_to_notify_after_sync> storage_service::sync_raft_t
             break;
         case node_state::rebuilding:
             // Rebuilding node is normal
-            co_await process_normal_node(id, rs);
+            co_await process_normal_node(id, host_id, ip, rs);
             break;
         default:
             on_fatal_internal_error(rtlogger, ::format("Unexpected state {} for node {}", rs.state, id));
         }
     };
 
+    sys_ks_futures.reserve(t.left_nodes.size() + t.normal_nodes.size() + t.transition_nodes.size());
+
     if (target_node) {
         raft::server_id raft_id{target_node->uuid()};
+        auto ip = _address_map.get(*target_node);
         if (t.left_nodes.contains(raft_id)) {
-            co_await process_left_node(raft_id);
+            co_await process_left_node(raft_id, *target_node, ip);
         } else if (auto it = t.normal_nodes.find(raft_id); it != t.normal_nodes.end()) {
-            co_await process_normal_node(raft_id, it->second);
+            co_await process_normal_node(raft_id, *target_node, ip, it->second);
         } else if ((it = t.transition_nodes.find(raft_id)) != t.transition_nodes.end()) {
-            co_await process_transition_node(raft_id, it->second);
+            co_await process_transition_node(raft_id, *target_node, ip, it->second);
         }
+        sys_ks_futures.push_back(raft_topology_update_ip(*target_node, ip, &nodes_to_notify));
     } else {
-        sys_ks_futures.reserve(t.left_nodes.size() + t.normal_nodes.size() + t.transition_nodes.size());
         for (const auto& id: t.left_nodes) {
-            co_await process_left_node(id);
+            locator::host_id host_id{id.uuid()};
+            auto ip = _address_map.find(host_id);
+            co_await process_left_node(id, host_id, ip);
+            if (ip) {
+                sys_ks_futures.push_back(raft_topology_update_ip(host_id, *ip, nullptr));
```

Comment:
Didn't previous version of the code notify in this case?
It would call `process_left_node(id)` which would call `remove_ip(..., true)` (`notify = true`) if IP was found (which in this case, it was)

Reply:
Hmm, maybe even if it did, it was a mistake; would we notify about the same set of left nodes every time we refreshed topology?

Or perhaps it would eventually stop once IP is gone. But then I wonder if there was a chance that we never sent any notification.


The correct thing to do would be probably to ensure that exactly once notification is sent, but I dunno if that can be done.

We need better tests for notifications...

Anyway perhaps here it should depend on whether `prev_normal` contained the node, I'm not sure

Reply:
process_left_node did not change. It still notifies about left node. raft_topology_update_ip only notifies about joining nodes.

Reply:
Ahhh ok. I got confused.
Would be more clear then if it took only joined set pointer, not the structure which contains both joined and left.
But it's a minor thing, not a blocker